content/blog/2019-09-25-vim-syntax-highlighting.md

```diff
@@ -0,0 +1,145 @@
+
+
+#### Limitations of Vimscript
+Despite its powerful features for syntax highlighting, Vimscript itself is rife with language issues that severely hinder its reliability and maintainability.
+
+References to old variable bindings remain active between executions of a script, a fact I often only realized after restarting Vim.
+
+Regular expressions are powerful, but possibly too powerful; for instance, the ``\ze`` regex atomdrops all matches regexes on both sides, keeping only the match on the left one. Such a regex is easy to express, but is computationally inefficient, potentially encouraging poor implementations.
+
+Finally, [metaprogramming](http://learnvimscriptthehardway.stevelosh.com/chapters/28.html) is idiomatic in Vimscript. 
+This is useful, since it allows us to factor out and reuse regexes from syntax group definitions into meta-strings. 
+However, Vimscript since makes a distinction between *strings* and *literal strings*, reasoning about "meta-strings" and "meta-literal-strings" can lead to hard to detect bugs.
+
+Semantically, a Vimscript string consists of a list of characters and characters 'escaped' by a backslash. (hence "\" is an invalid string; it is a malformed character escape).
+On the other hand, a literal string simply consists of the literal characters in the string.
+If we attempt to factor out a regexp as a string, any escaped characters will be stored as 'escaped', rather than as characters.
+Splicing this faulty regexp into a program string will generate a faulty program string containing escaped characters.
+
+We can solve this by *escaping each 'escaping' backslash with a backslash, except for those backslashes that escape backlashes.* (it's no wonder the leading Vimscript tutorial recommends [beer](http://learnvimscriptthehardway.stevelosh.com/chapters/08.html) to accompany some of the exercises.)
+To avoid backslash hell, we can instead store regexps as literal strings, which can never escape characters.
+Later on, these can be spliced with program fragments that are also represented as string literals, rather than strings, to avoid escaping any hard-coded regexps.
+
+Once I dealt with the frustrations and mysterious bugs, I found Vimscript to be pretty fun, in part due to its flexibility,
+and I enjoyed pondering the feedback problem of language design for language tooling.
+
+### Design Considerations and Future Work
+As I've described it, a syntax highlighters principle job is to give a meaningful visual presentation to an evolving syntax tree.
+In reality, developers have come to expect much more from a syntax highlighter, including:
+- Syntactically-aware code editing and navigation, such as automatic indentation, folding, variable renaming, etc...
```

Comment:
```suggestion
- Syntactically-aware code editing and navigation, such as automatic indentation, folding, variable renaming, etc.
```

I think "etc." is fine by itself; it doesn't need to be doubled up with an ellipsis.